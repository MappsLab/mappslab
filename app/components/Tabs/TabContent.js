// @flow
import * as React from 'react'
import { TabConsumer } from './Tabs'
import type { TabContextValue } from './Tabs'

/**
 * TabContent
 */

type BaseProps = {
	children: React.Node,
	displayName?: void | string,
	name: string,
}

type Props = BaseProps & TabContextValue

class TabContent extends React.Component<Props> {
	static defaultProps = {
		displayName: undefined,
	}

	componentDidMount() {
		const { registerTab, name, displayName } = this.props
		registerTab({ name, displayName })
	}

	componentWillUnmount() {
		const { removeTab, name } = this.props
		removeTab({ name })
	}

	render() {
		const { name, children, activeTab } = this.props
		return activeTab && name === activeTab.name ? children : null
	}
}

const Wrapper = (props: BaseProps) => <TabConsumer>{(contextValue) => <TabContent {...contextValue} {...props} />}</TabConsumer>

Wrapper.defaultProps = {
	displayName: undefined,
}

export default Wrapper
