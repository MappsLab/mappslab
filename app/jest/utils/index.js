export * from './rendering'
export * from './mockServer'
