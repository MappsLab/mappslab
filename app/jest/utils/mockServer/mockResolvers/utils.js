import * as R from 'ramda'
import faker from 'faker'

export const paginate = (nodes) => ({
	pageInfo: {
		hasNextPage: false, // todo, update this to actually work
		hasPrevPage: false,
	},
	edges: nodes.map((node) => ({
		cursor: node.id,
		node,
	})),
})

export const generateUid = () => `0x${faker.random.uuid().substr(0, 8)}`

export const extractUid = (parent, args) => R.path(['cursor'])(parent) || R.path(['input', 'uid'])(args) || generateUid()

export const generateRandomArray = (arr) => faker.helpers.shuffle(arr).slice(0, 3)

export const slugify = (text) =>
	text
		.toString()
		.toLowerCase()
		.replace(/\s+/g, '-')
		.replace(/[^\w-]+/g, '')
		.replace(/--+/g, '-')
		.replace(/^-+/, '')
		.replace(/-+$/, '')

const capitalize = (str) => {
	const [first, ...rest] = str.split('')
	return `${first.toUpperCase()}${rest.join('')}`
}

export const deSlugify = (text) =>
	text
		.split('-')
		.map(capitalize)
		.join(' ')

export const randomAmount = ({ max, min }: { max: number, min: number }, cb: Function): Array<mixed> => {
	if (!max) throw new Error('randomAmount({ max }): max has to be defined!')
	const n = faker.random.number({ min: min || 0, max })
	const result = []
	for (let i = 0; i < n; i += 1) {
		result.push(cb())
	}
	return result
}

export const probably = (positive: any, negative: any = null): any => (faker.random.number(100) > 25 ? positive : negative)
export const probablyNot = (positive: any, negative: any = null): any => probably(negative, positive)

export const mostLikely = (positive: any, negative: any = null): any => (faker.random.number(100) > 5 ? positive : negative)
export const mostLikelyNot = (positive: any, negative: any = null): any => mostLikely(negative, positive)
