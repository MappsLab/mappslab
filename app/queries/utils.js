// @flow
import type { ComponentType } from 'react'
import * as R from 'ramda'
import { graphql } from 'react-apollo'
// import { mapProps } from 'recompose'

// export const defaultPropsToMap = {
// 	props: ({ ownProps, data }) => {
// 		return {
// 			...data,
// 		}
// 	},
// }

const _makeQuery = (query, config: Object, Component: ComponentType<any>) => graphql(query, config)(Component)

export const makeQuery = R.curry(_makeQuery)

export const unwindEdges = (prop: string, data: Object): Object | Array<Object> => {
	return R.pipe(
		R.when(
			R.path([prop, 'edges']),

			R.pipe(
				// Get the prop -> edges
				R.path([prop, 'edges']),
				// Get all of the nodes
				R.pluck('node'),
				// Check each node to see if there are more edges to unwind
				R.map(
					// For each key of the node,
					R.mapObjIndexed(
						(value, key, obj) =>
							value.edges
								? // If the key is 'edge', unwind it
								  unwindEdges(key, obj)
								: // otherwise, just return the value
								  value,
					),
				),
			),
		),
	)(data)
}
