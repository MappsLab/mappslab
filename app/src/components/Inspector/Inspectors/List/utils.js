// @flow
import type { ClassroomType, UserType, MapType } from 'Types'
// import type { InspectorItem } from '../../InspectorProvider'

type NodeArray = Array<ClassroomType | UserType | MapType>

export type SearchForList = (any) => Promise<NodeArray> | NodeArray

export type ListItemHandler = (ClassroomType | UserType | MapType) => void

export type ListItem = {
	node: ClassroomType | UserType | MapType,
	key: string,
	onClick: () => void,
}

// const nodeToInspectorItem = (item: ClassroomType | UserType | MapType): InspectorItem => ({
// 	uid: item.uid,
// 	type: item.__typename,
// 	title: item.__typename === 'User' ? item.name : item.title,
// })

export const nodeToListItem = (node: ClassroomType | UserType | MapType, handler: ListItemHandler): ListItem => {
	// $FlowFixMe TODO: make this detect types
	const { uid } = node
	return {
		key: uid,
		node,
		info: [],
		onClick: () => handler(node),
	}
}
