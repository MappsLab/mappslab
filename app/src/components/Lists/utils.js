// @flow
import type { Node, Variables } from 'Types'
// import type { InspectorItem } from '../../InspectorProvider'

// type NodeArray = Array<ClassroomType | UserType | MapType>

export type SearchForList = (any) => Promise<void> | void

export type ListItemHandler = (Node) => void

export type ListItem = {
	node: Node,
	onClick: () => void,
}

export type ListOfTypeBaseProps<Type> = {
	title: string,
	items: Array<Type>,
	viewerCanAdd: boolean,
	update: (Type) => void,
	onItemClick: ListItemHandler,
}

export type ListOfTypeProps<Type> = ListOfTypeBaseProps<Type> & {
	searchQuery: (Variables) => Promise<void>,
	searchResults: Array<Type>,
}

export const nodeToListItem = (node: Node, handler: ListItemHandler): ListItem => {
	// $FlowFixMe TODO: make this detect types
	const { uid } = node
	return {
		key: uid,
		node,
		info: [],
		onClick: () => handler(node),
	}
}
