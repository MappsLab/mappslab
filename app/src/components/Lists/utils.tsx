import { NodeType } from '../../types-ts'

export type SearchForList = (arg0: any) => Promise<void> | void

export type ListItemHandler<T extends NodeType> = (
	item: T,
) => void | Promise<void>

export type CreateNewFn = (str: string) => void | Promise<void>

export interface ListItem {
	key: string
	node: NodeType
	info: Array<string>
	onClick: () => void | Promise<void>
}

export interface ListOfTypeProps<Type extends NodeType> {
	title: string
	items?: Type[]
	viewerCanAdd: boolean
	update?: (node: NodeType) => void | Promise<void>
	onItemClick?: ListItemHandler<Type>
	onSearchResultClick?: ListItemHandler<Type>
	create?: (args: any) => Promise<void> | void
	remove?: (args: any) => Promise<void> | void
}

export const nodeToListItem = <T extends NodeType>(
	node: T,
	handler: ListItemHandler<T>,
): ListItem => {
	const { uid } = node
	return {
		key: uid,
		node,
		info: [],
		onClick: () => handler(node),
	}
}
