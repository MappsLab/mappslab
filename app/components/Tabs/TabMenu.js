// @flow
import React from 'react'
import styled from 'styled-components'
import { TabConsumer } from './Tabs'
import type { TabContextValue } from './Tabs'

const TabButton = styled.button`
	${({ theme, active }) => `
		font-size: ${theme.text.h3};
		border-bottom: 2px solid ${active ? 'black' : 'transparent'};
	`};
`

/**
 * TabMenu
 */

const TabMenu = () => (
	<TabConsumer>
		{({ tabs, changeTab, activeTab }: TabContextValue) => {
			if (!tabs) return null
			return tabs.map(({ name, displayName }) => (
				<TabButton key={name} active={name === activeTab} onClick={changeTab({ name })}>
					{displayName || name}
				</TabButton>
			))
		}}
	</TabConsumer>
)

export default TabMenu
