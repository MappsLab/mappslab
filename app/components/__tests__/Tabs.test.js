/* eslint-disable no-undef */
import React from 'react'
import { fireEvent } from 'react-testing-library'
import { render } from 'Jest/utils'

import Tabs from '../Tabs'

const normalTabs = (
	<Tabs initialTab="Nikes">
		<Tabs.Menu />
		<div data-testid="tabContainer">
			<Tabs.Tab name="Nikes">These bitches want Nikes</Tabs.Tab>
			<Tabs.Tab name="Ivy">I thought that I was dreaming</Tabs.Tab>
			<Tabs.Tab name="Pink + White">Yeah, yeah oh</Tabs.Tab>
		</div>
	</Tabs>
)

describe('Tabs Component', () => {
	it('Mounts Correctly', () => {
		/* Arrange */
		const { container } = render(normalTabs)
		/* Assert */
		expect(container).toMatchSnapshot()
	})

	it('should display a button for each tab', async () => {
		const { getByTestId } = render(normalTabs)
		expect(getByTestId('tabButton-Nikes')).toHaveTextContent('Nikes')
		expect(getByTestId('tabButton-Ivy')).toHaveTextContent('Ivy')
		expect(getByTestId('tabButton-Pink + White')).toHaveTextContent('Pink + White')
	})

	it('should only display one tab at a time', async () => {
		const { getByTestId } = render(normalTabs)
		expect(getByTestId('tabContainer')).toHaveTextContent('These bitches want Nikes')
		expect(getByTestId('tabContainer')).not.toHaveTextContent('I thought that I was dreaming')
	})

	it('should switch tabs when the TabMenu item is clicked', async () => {
		const { getByTestId } = render(normalTabs)
		const tab2button = getByTestId('tabButton-Ivy')
		fireEvent.click(tab2button)

		expect(getByTestId('tabContainer')).not.toHaveTextContent('These bitches want Nikes')
		expect(getByTestId('tabContainer')).toHaveTextContent('I thought that I was dreaming')
	})

	it('should throw an error if two tabs have the same name', async () => {
		const renderTabs = () =>
			render(
				<Tabs>
					<Tabs.Menu />
					<Tabs.Tab name="bad">bad!</Tabs.Tab>
					<Tabs.Tab name="bad">BAD!</Tabs.Tab>
				</Tabs>,
			)
		expect(renderTabs).toThrow('bad is not valid: all tab names must be unique')
	})

	it.skip('should remove the tab header when corresponding Tab unmounts', async () => {
		/* Arrange */
		// const { container, getByTestId } = render( ... )
		/* Act */
		/* Assert */
		// expect(...)
	})
})
