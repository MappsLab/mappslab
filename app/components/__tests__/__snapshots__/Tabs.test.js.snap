// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Component Mounts Correctly 1`] = `
.c0 {
  font-size: 16px;
  border-bottom: 2px solid transparent;
}

<div>
  <button
    class="c0"
    data-testid="tabButton-Nikes"
  >
    Nikes
  </button>
  <button
    class="c0"
    data-testid="tabButton-Ivy"
  >
    Ivy
  </button>
  <button
    class="c0"
    data-testid="tabButton-Pink + White"
  >
    Pink + White
  </button>
  <div
    data-testid="tabContainer"
  >
    These bitches want Nikes
  </div>
</div>
`;
